blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/BlueOceanCredentialsProvider.java

```diff
@@ -0,0 +1,262 @@
+    }
+
+    public static class FolderPropertyImpl extends AbstractFolderProperty<AbstractFolder<TopLevelItem>> {
+        private final String user;
+        private final String id;
+
+        @DataBoundConstructor
+        public FolderPropertyImpl(@Nonnull String user, @Nonnull String id) {
+            this.user = user;
+            this.id = id;
+        }
+
+        public @Nonnull String getUser() {
+            return user;
+        }
+
+        public @Nonnull String getId() {
+            return id;
+        }
+
+        @Override
+        public AbstractFolderProperty<?> reconfigure(StaplerRequest req, JSONObject form) throws FormException {
+            return this;
+        }
+
+        @Extension(optional = true)
+        public static class DescriptorImpl extends AbstractFolderPropertyDescriptor {
+        }
+
+        public static class CredentialsStoreActionImpl extends CredentialsStoreAction {
+
+            private final CredentialsStore store;
+
+            CredentialsStoreActionImpl(CredentialsStore store) {
+                this.store = store;
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Nonnull
+            @Override
+            public CredentialsStore getStore() {
+                return store;
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public String getIconFileName() {
+                return isVisible()
+                        ? "/plugin/credentials/images/48x48/folder-store.png"
+                        : null;
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public String getIconClassName() {
+                return isVisible()
+                        ? "icon-credentials-folder-store"
+                        : null;
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public String getDisplayName() {
+                return DISPLAY_NAME;
+            }
+
+            @Override
+            public String getUrlName() {
+                return super.getUrlName();
+            }
+        }
+
+        public static class StoreImpl extends CredentialsStore{
+            private final AbstractFolder abstractFolder;
+            private final CredentialsStoreAction storeAction;
+
+            StoreImpl(AbstractFolder abstractFolder) {
+                super(BlueOceanCredentialsProvider.class);
+                this.abstractFolder = abstractFolder;
+                this.storeAction = new CredentialsStoreActionImpl(this);
+            }
+
+            @Nonnull
+            @Override
+            public List<Domain> getDomains() {
+                //XXX: how to get domain from given Credentials object? Maybe attach domain name to AbstractFolder?
+                return Collections.emptyList();
```

Comment:
actually what I see now is that you want something different, you want

`Collections.singletonList(/*domain for "blueocean-github-domain"*/)`

The domain is about matching the URL the credentials are requested for against the URLs that the credentials are allowed to be used against.

Scope is a different cross-cutting requirement